avocado/core/requirements/resolver.py

```diff
@@ -0,0 +1,30 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2021
+# Authors: Willian Rampazzo <willianr@redhat.com>
+
+from ..nrunner import Runnable
+
+
+class RequirementsResolver:
+
+    name = 'requirements'
+    description = 'Requirements resolver for tests with requirements'
+
+    @staticmethod
+    def resolve(runnable):
+        requirements_runnables = []
+        for requirement in runnable.requirements:
+            kind = 'requirement-%s' % requirement.get('type')
```

Comment:
Shouldn't this be a `requirement.pop()` instead? Wouldn't this keep `type` as a kwargs?

Reply:
So, some version ago, you made this comment https://github.com/avocado-framework/avocado/pull/4481#discussion_r603426478 about removing it from the dictionary. That was a copy, I know, but at the time, I agreed with you. The runner will handle the kwargs and not use this if it is not necessary.

Thinking again about it, the runner knows its `type`, so I don't see harming if I pop it from the dictionary.

Reply:
Yeah, I think my main point then was not needing to edit the dictionary (removing one entry) if it was a copy that would be discarded.

My main point *now* is that we don't want to pass extra arguments (such as `type`) to the `kwargs` which are not accounted for, for correctness reasons.

Having said that, my suggestion of `pop`ing the `type` *may* actually be incorrect, because we'd be changing the individual `runnable.requirement` entries.  **If there are no other side effects**, I think it's fair game to say that after a resolution is done, the requirements may be changed.  To be extra safe at this time, though, we may way to keep things simple and **not** modify the individual `runnable.requirements` though.